### PR TITLE
[Multi-sig] feat: implement simple payment multisig (and voting via multiple roles)

### LINKED ISSUE #2188: [Multi-sig] feat: implement simple payment multi-sig motion
### Story

As a *user moderating a colony*, I want to be able to create and finalize a multi-sig simple payment motion.

### Description

After the groundwork has been laid with #2150 , implement the simple payment motion.

~~The `voteOnMultiSig` saga will need adjusting to use multicall if a user has both funding and administration permissions and should call `changeVote` twice with the different permissions.~~ Try this without multicall first as the contracts might handle this scenario for us

Please also check the approval / rejection signatures work as intended as this is the only action requiring two permissions (although hopefully this has already been covered by #2543)

### This issue is complete when...

- [ ] The simple payment multi-sig motion can be created
- [ ] The `voteOnMultiSig` uses multicall when necessary
- [ ] The motion can be finalized
- [ ] Approvals / rejections are accurately counted
- [ ] Only users with [adequate permissions](https://www.notion.so/colony/Multi-sig-Permissions-6aaf0b95309f4754a243d54e8d9c4189?pvs=4#fec3699a5fd241a2925bd2809494b157) can create them

### PR DESCRIPTION
**Pull request by bassgeta**
## Description

[Block ingestor PR](https://github.com/JoinColony/block-ingestor/pull/248/)
This PR adds support for simple payments motion, but also adds support for multiple votes per user - they vote for each role they hold.

## Testing

This one's gonna be really fun! :melting_face: I'll try to keep it as simple as it gets, but you will need 3 browsers.
Browser A -> `leela`
Browser B -> `amy`
Browser C -> `fry`

Query to get signatures on a multisig:
```
query MyQuery {
  getColonyAction(id: "YOUR_TX_HASH") {
    id
    isMultiSig
    multiSigData {
      isExecuted
      isRejected
      multiSigDomainId
      signatures {
        items {
          colonyAddress
          multiSigId
          role
          vote
          user {
            id
            profile {
              displayName
            }
          }
        }
      }
    }
    type
  }
}
```

The prerequisite for testing all of these are the following steps
1. As `Leela` install the multi-sig extension (don't forget to enable permissions!) 
2. Set it's threshold to `Fixed` and a value of 2 (then refresh the page)
image
3. Assign the `Payer` multi-sig role to `amy` in team `General`
image
4. Assign `fry` 1 or 2 of the following roles `Administration`, `Funding`, `Arbitration`, just don't do all 3
image

After all this is done, refresh the page and let's go!

### Case 1: Sign as Owner and single role holder
1. As `leela` create a `Simple payment` action with `Multi-sig` as the decision method and the team should be `General`
2. After it's created you need to see 3 out of 6 votes, verify it by running the query above (`leela` voted with all 3 required roles)
image
image
3. As `amy` open the action in another browser and Approve it. The action should go to the `Finalize` step, but if you click on `Approval` you should be able to see that both users voted with 3 roles. Verify by running the query.
image
4. Finalize as any of these two (or even `fry`), after refreshing `fry`'s browser, their CREDS balance should increase by 5
image
image

### Case 2: Create the multisig motion as a `Payer`
I don't wanna bloat the description, but test case 1, just instead of creating the motion with `leela` do it with `amy`, everything should still work :+1: 

### Case 3: Sign as Owner and `fry`
1. As `leela` create a `Simple payment` action with `Multi-sig` as the decision method and the team should be `General`
2. After it's created you need to see 3 out of 6 votes, verify it by running the query above (`leela` voted with all 3 required roles)
image
image
3. Open the action as `fry` in another browser and Approve it. Their vote should just count as 1 towards the progress.
image
image
image
4. Open the action as `amy`, and approve it. When going back to the `Approval` step, you should see 7/6 votes
image
5. Finalize it and verify that `fry` has indeed received the CREDS

### Case 4: Reject a motion as Owner and Payer
This is similar to case 1, just do the same steps, but instead of voting for it, `Reject` it
1. Create it as `leela`
2. Remove your vote and vote for `Reject`
image
3. As `amy` also reject the motion, you should be able to `Finalize cancel` (click on the `Approval` step to see the votes)
image
4. `Finalize cancel` and poor `fry` is without his extra CREDS
image

### Case 5. Reject a motion as Owner and `fry`
This is similar to case 3, just do the same steps, but instead of voting for it, `Reject` it

1. Create it as `leela`
2. Remove your vote and vote for `Reject`
image
3. As `fry` reject the motion, the progress should update to 4/6
image
4. As `amy` reject it and `Finalize cancel`. `fry` should not get the CREDS
image

Now if you really wanna put this through its paces continue!

### Case 5: Sign it as `fry` in Andromeda 
Before doing this, take a note of how many funds are in `Andromeda`
image

1. Remove `fry`'s permissions in `General`, just give him the `Architecture` multi-sig role, it shall suffice for this test
image
2. Assign `fry` `Administration` and `Funding` multi-sig roles in team `Andromeda`
image
3. Create a simple payment action as `leela`, then also sign it as `fry`.
image
4. Finally sign this as `amy`, check `fry`'s balance, they should get 5 CREDS, and the balance in team `Andromeda` should drop
image

At this point, you probably get the just of it, so you can also try rejecting in a subdomain, etc etc, the description is getting too long :)

## Diffs

**New stuff** ✨

* `RoleApprovalTooltip` component which displays the number of roles user voted with and the tooltip
* simple payment multisig support

**Changes** 🏗

* Block ingestor now allows duplicate votes if they are made by different roles
* MultiSigWidget's `ApprovalStep` now aggregates all the votes per user and has an array of roles
* payment motion now also has multisig support
* Temporary filter for user's role voting to only vote with the roles they have, not try to vote with all required roles (getMultiPermissionProofs would throw an error), will be reworked when we merge [this PR](https://github.com/JoinColony/colonyCDapp/pull/2674)

Resolves #2188

### CODE REVIEW

**Review by mmioana — COMMENTED** [review]
Amazing job @bassgeta 👏 

And really cool testing steps
Screenshot 2024-07-10 at 13 47 37
Screenshot 2024-07-10 at 13 48 32
Screenshot 2024-07-10 at 13 50 05
Screenshot 2024-07-10 at 13 50 39
Screenshot 2024-07-10 at 13 53 19
Screenshot 2024-07-10 at 14 10 30
Screenshot 2024-07-10 at 14 11 50


The only thing that I have noticed is that Fry has 3 votes shown in the pill before approving/rejecting an action. Is this as intended?

Screenshot 2024-07-10 at 14 16 41

**Comment by bassgeta** [comment]
> Amazing job @bassgeta 👏
> 
> And really cool testing steps Screenshot 2024-07-10 at 13 47 37 Screenshot 2024-07-10 at 13 48 32 Screenshot 2024-07-10 at 13 50 05 Screenshot 2024-07-10 at 13 50 39 Screenshot 2024-07-10 at 13 53 19 Screenshot 2024-07-10 at 14 10 30 Screenshot 2024-07-10 at 14 11 50
> 
> The only thing that I have noticed is that Fry has 3 votes shown in the pill before approving/rejecting an action. Is this as intended?
> Screenshot 2024-07-10 at 14 16 41

Ah I forgot to note, the `Awaiting 3` part isn't working here, so I just have a placeholder in place, we need to rework how we fetch eligible signees, and then pass the roles they're holding into this new component. I felt it was a bit out of scope, so I just stubbed it in this PR

**Comment by mmioana** [comment]
@bassgeta Also I have noticed the action type appears as generic in the sidebar, I think en-action.ts needs to be updated
Screenshot 2024-07-10 at 14 26 57

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great work on this so far @bassgeta 🎉 Some great work in here. Nice job adding the role to the voting.

I followed the nice and detailed testing steps:

Screenshot 2024-07-12 at 08 27 08

Screenshot 2024-07-12 at 08 27 43

Screenshot 2024-07-12 at 08 32 11

Screenshot 2024-07-12 at 08 44 03

Screenshot 2024-07-12 at 08 44 18

 I've found one tiny issue with the tooltip.
 
Screenshot 2024-07-12 at 08 30 38

I've hit a few issues with the permissions notice when a user only has multi-sig permissions in a sub-domain. At one point the multi-sig decision method even stopped showing all together.

Screenshot 2024-07-12 at 08 37 28 

https://github.com/user-attachments/assets/6ec40b10-000f-4eaf-9063-f3e6db86fb94

The notice seems to be checking the `from` field rather than the `createdIn` field.

https://github.com/user-attachments/assets/5e54a7f0-9e67-4c09-b7a4-90c0b61bca3d

I think it is accurate that the contracts require you to have all 3 permissions in the same domain to create the multi-sig motion, but the permissions notice should catch this before the saga.

https://github.com/user-attachments/assets/2314eb60-fb38-4c25-9fc7-095fc2f040eb

Also noting the same `Generic` action issue, but also the `Created in` field is missing.
 
Screenshot 2024-07-12 at 08 44 56

I think if the permissions notice issues are only affecting this action, these could be addressed in a separate issue as to not hold this one up, but if it is going to cause issues elsewhere better to resolve them here. (Should be able to tell more easily after a rebase now that some other actions are in the feature branch! 🎉 )

Unfortunately seem to be some issues introduced to the reputation motions. I tried creating one as a user with no permissions.

https://github.com/user-attachments/assets/ad691db3-8246-415a-b108-3594d688e1b2

And then had a different issue as a user with permissions:

https://github.com/user-attachments/assets/aba6ce2c-fff4-418d-8166-4453a9e42ea9

Overall though some great work, nice to see the multiple roles voting working!

**Comment by iamsamgibbs** [comment]
I'm running into issues with the permissions error on another branch, so I don't think any of those are specific to this PR.

**Review by davecreaser — APPROVED** [review]
Very nice work! This was for sure a complicated one 😄 I've just left a couple of tiny comments, not super important things though.

Also tested with your instructions and all looks good! Here's a bunch of screenshots from that:
Screenshot 2024-07-17 at 11 22 25
Screenshot 2024-07-17 at 11 22 56
Screenshot 2024-07-17 at 11 24 17
Screenshot 2024-07-17 at 11 25 28
Screenshot 2024-07-17 at 11 26 43
Screenshot 2024-07-17 at 11 27 19
Screenshot 2024-07-17 at 11 28 58
Screenshot 2024-07-17 at 11 30 11
Screenshot 2024-07-17 at 11 32 26
Screenshot 2024-07-17 at 11 35 09

File: src/redux/sagas/motions/paymentMotion.ts [hunk 1]
```diff
@@ -163,28 +107,199 @@ function* createPaymentMotion({
...
+             * This will make it so that the user only receives reputation in the
+             * above domain, but none in the skill itself.
+             */
+            0,
+          ],
+        );
+
+        return {
+          context: ClientType.MultisigPermissionsClient,
+          methodName: 'createMotion',
```
Comment (iamsamgibbs) [inline src/redux/sagas/motions/paymentMotion.ts:203]: Could be `TRANSACTION_METHODS.CreateMotion`

File: src/redux/sagas/motions/paymentMotion.ts [hunk 2]
```diff
@@ -163,28 +107,199 @@ function* createPaymentMotion({
...
+      );
+
+      const { key, value, branchMask, siblings } = yield call(
+        colonyClient.getReputation,
+        skillId,
+        AddressZero,
+      );
+      return {
+        context: ClientType.VotingReputationClient,
+        methodName: 'createMotion',
```
Comment (iamsamgibbs) [inline src/redux/sagas/motions/paymentMotion.ts:279]: Could be `TRANSACTION_METHODS.CreateMotion`

File: src/redux/sagas/motions/paymentMotion.ts [hunk 3]
```diff
@@ -163,28 +107,199 @@ function* createPaymentMotion({
...
+          id: metaId,
+          index: 0,
+        },
+        ready: false,
+      };
+    }
+
+    const transactionParams = yield getPaymentMotionParams();
+    // create transactions
+    yield fork(createTransaction, createMotion.id, transactionParams);
```
Comment (iamsamgibbs) [inline src/redux/sagas/motions/paymentMotion.ts:302]: Could be `TRANSACTION_METHODS.AnnotateTransaction`

**Review by iamsamgibbs — APPROVED** [review]
Great work on this! All working nicely for me. Just the tiny issues Dave already raised.

Both issues I raised previous are resolved:

Screenshot 2024-07-17 at 15 21 39

Screenshot 2024-07-17 at 15 22 16

I tested all the steps again to ensure nothing has broken:

Screenshot 2024-07-17 at 15 17 55

Screenshot 2024-07-17 at 15 20 22

Screenshot 2024-07-17 at 15 21 08

**Review by mmioana — APPROVED** [review]
Amazing work @bassgeta and really throughout testing steps! 🚀 

Screenshot 2024-07-17 at 21 14 14
Screenshot 2024-07-17 at 21 14 53
Screenshot 2024-07-17 at 21 16 08
Screenshot 2024-07-17 at 21 16 55
Screenshot 2024-07-17 at 21 17 29
Screenshot 2024-07-17 at 21 17 40
Screenshot 2024-07-17 at 21 18 03
Screenshot 2024-07-17 at 21 19 05
Screenshot 2024-07-17 at 21 19 37
Screenshot 2024-07-17 at 21 20 02
Screenshot 2024-07-17 at 21 20 43
Screenshot 2024-07-17 at 21 21 12
Screenshot 2024-07-17 at 21 23 09
Screenshot 2024-07-17 at 21 23 29
Screenshot 2024-07-17 at 21 29 32

**Pull request merged**